### PR TITLE
Update server and make it run from cache directory

### PR DESCRIPTION
**Pull request by rchl**
- update `vscode-json-languageserver` to version 1.2.3
 - update schemas.py with latest updates from http://schemastore.org/api/json/catalog.json
 - add palette command for opening LSP-json settings
 - run server from cache directory

Of course will wait with integrating until `lsp_utils` is released.

### CODE REVIEW

**Comment by rchl** [comment]
I would also flatten options as I did for LSP-eslint to allow for overriding only individual options and make it easier to understand.

**Comment by rchl** [comment]
Since I jsut saw this https://github.com/sublimelsp/LSP-json/pull/6/files#diff-0f49ebf0103c2d85a2b61d6712cad9c3 , I guess we should better support `languages` in an array form.

File: plugin.py [hunk 1]
```diff
@@ -1,99 +1,97 @@
...
+    def migrate_and_read_configuration(self) -> None:
+        settings = {}
+        loaded_settings = sublime.load_settings(SETTINGS_FILENAME)
+
+        if loaded_settings:
+            if loaded_settings.has('client'):
+                client = loaded_settings.get('client')
+                loaded_settings.erase('client')
+                # Migrate old keys
+                for key in ['languages', 'initializationOptions', 'settings']:
```
Comment (predragnikolic) [inline plugin.py]:
```
                # Migrate old keys
                for key in client:
                    loaded_settings.set(key, client[key])
                sublime.save_settings(SETTINGS_FILENAME)
```
This here can be just a bit simplified... a bit :)

Comment (rchl) [reply plugin.py]: Thanks. Applied.
I also noticed there was a bug with not reading `languages` key in the code below.

**Comment by jfcherng** [comment]
`vscode-json-languageserver` marks trailing commas as `warning` since `1.2.2` ([this commit](https://github.com/vscode-langservers/vscode-json-languageserver/blob/67425469237648c2c849276a4852e74906296d29/src/jsonServerMain.ts#L376)) and it looks like it's hard-coded. I wonder there is a way to "ignore" this (for ST's config files, at least).

image

**Comment by rchl** [comment]
But then the latest VSCode doesn't report trailing commas in its configuration files. Should probably investigate why is that.

**Comment by jfcherng** [comment]
- I see a suspicious commit: https://github.com/vscode-langservers/vscode-json-languageserver/commit/c87633b3923ef7ed336f3285cb7b73fd0bf99e8a
  But there is actually no code change. So it looks like it's related to its dependency: `vscode-json-languageservice`
- I then check `vscode-json-languageservice`'s [changelog ](https://github.com/microsoft/vscode-json-languageservice/blob/0947a45770104ddcbb22f5d88951f6582b8e102f/CHANGELOG.md#334--2019-09-20). Notice that in `3.3.3`: `Schemas can configure whether comments and/or trailing commas are permitted.`
- So it looks like it's related to JSON scheme: https://github.com/microsoft/vscode-json-languageservice/blob/289550209598e45e5c483f8e71a9bab262fb486d/src/jsonSchema.ts#L67

**Comment by rchl** [comment]
Thanks for that. I've looked further from here and came up with this:
 - vscode registers internal schemas and exposes them as URLs. For example `vscode://schemas/keybindings` for keybindings [1]
 - LSP client (vscode) sends a `json/schemaAssociations` notification to json server with filePattern => schema mappings:
```
Sending notification 'json/schemaAssociations'.
Params: {
    "vscode://defaultsettings/keybindings.json": [
        "vscode://schemas/keybindings"
    ],
    "/User/keybindings.json": [
        "vscode://schemas/keybindings"
    ],
    ...
```
 - json server resolves the schema by loading and parsing it and uses when validating file

Not sure how we could do something similar for Sublime since URL of the schema should ideally be resolvable from a local machine rather than the Internet. If we make it resolve from the Internet then it would be easier (all schemas [already included in LSP-json](https://github.com/sublimelsp/LSP-json/blob/master/schemas.py) are resolved from the Internet already).


[1] https://github.com/microsoft/vscode/blob/839383dc1d843d8dd94823c5f1c99da7500bc1ee/src/vs/workbench/services/keybinding/browser/keybindingService.ts#L678-L727

**Comment by rchl** [comment]
And as far as I can see in the code, the trick with the schema will work regardless if `languageId` is `json` or `jsonc`. So nothing would be needed from SublimeLSP side.

**Comment by rchl** [comment]
Actually readme of the server describes this pretty well - https://github.com/vscode-langservers/vscode-json-languageserver

What protocols server fetches itself can be configured with a `handledSchemaProtocols` option. Schemas that it can't handle (like `vscode://`, it will request from the client through `vscode/content` message).

**Comment by rchl** [comment]
Added script to update schema list and replaced schemas.py with schemas.json that is loaded once on first retrieve of configuration.

**Comment by rchl** [comment]
Went all-in and added schemes for SublimeText-specific files. All with the exception of `sublime-keymap` schema contain bare-minimal to not complain about trailing commas and comments. The `sublime-keymap` schema also contains type definitions for things that are allowed (although not 100% complete).

File: plugin.py [hunk 2]
```diff
@@ -1,99 +1,88 @@
...
+        default_configuration.update(configuration)
+
+        if 'json' not in default_configuration['settings']:
+            default_configuration['settings']['json'] = {}
+        if not self._default_schemas:
+            schemas = ['schemas_extra.json', 'schemas.json']
+            for schema in schemas:
+                path = 'Packages/{}/{}'.format(PACKAGE_NAME, schema)
+                self._default_schemas.extend(json.loads(ResourcePath(path).read_text()))
+        default_configuration['settings']['json']['schemas'] = self._default_schemas
```
Comment (rwols) [inline plugin.py]: you can use `.setdefault` from the `dict` object here:

```python
default_configuration['settings'].setdefault('json', {})['schemas'] = self._default_schemas
```

this will add `'json': {}` if it doesn't exist yet, and will return the value if `'json'` does exist.

File: schemas_extra.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    "fileMatch": ["*.sublime-commands"],
+    "url": "sublime://schemas/sublime-commands"
+  },
+  {
+    "fileMatch": ["*.sublime-keymap"],
+    "url": "sublime://schemas/sublime-keymap"
+  },
+  {
+    "fileMatch": ["*.sublime-macro"],
+    "url": "sublime://schemas/sublime-macro"
```
Comment (rwols) [inline schemas_extra.json]: I really like this solution! It'll also allow us to make diagnostics more precise!

File: plugin.py [hunk 4]
```diff
@@ -1,99 +1,87 @@
...
+            schemas = ['schemas_extra.json', 'schemas.json']
+            for schema in schemas:
+                path = 'Packages/{}/{}'.format(PACKAGE_NAME, schema)
+                self._default_schemas.extend(json.loads(ResourcePath(path).read_text()))
+
+        default_configuration['settings'].setdefault('json', {})['schemas'] = self._default_schemas
+
+        return read_client_config(self.name, default_configuration)
+
+    def migrate_and_read_configuration(self) -> None:
```
Comment (jfcherng) [inline plugin.py]: the return type is `dict`

Comment (predragnikolic) [inline plugin.py]: The return type is not correct`-> None`
You probably meant a dict

File: plugin.py [hunk 5]
```diff
@@ -102,4 +90,11 @@ def on_start(self, window) -> bool:
...
     def on_initialized(self, client) -> None:
-        pass
\ No newline at end of file
+        client.on_request(
+            'vscode/content',
+            lambda params, request_id: self.handle_vscode_content(client, params, request_id))
+
+    def handle_vscode_content(self, client, uri, request_id):
+        name = uri.split('/')[-1]
+        schema_resource = ResourcePath('Packages', PACKAGE_NAME, 'schemas', '{}.json'.format(name))
```
Comment (predragnikolic) [inline plugin.py]: This is so far the thing I am the most excited about :)
I think that this could be improved in the future 
to allow other sublime plugins to create a schema for their settings.

File: schemas/jsconfig.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (predragnikolic) [inline schemas/jsconfig.json]: Is there any reason you included jsconfg.json and tsconfig.json in the schema folder?

File: schemas/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (predragnikolic) [inline schemas/tsconfig.json]: Is this file necessary? :)

File: schemas/vscode-base.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (predragnikolic) [inline schemas/vscode-base.json]: Also what is the purpose of this file? (tried googling it, but found nothing on the first glance) :)

File: schemas_extra.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    "fileMatch": ["/.vscode/*.json"],
+    "url": "sublime://schemas/vscode-base"
+  },
+  {
+    "fileMatch": ["/package.json"],
+    "url": "sublime://schemas/package-json-eslint"
+  },
+  {
+    "fileMatch": ["tsconfig.json", "tsconfig.*.json", "tsconfig-*.json"],
+    "url": "sublime://schemas/tsconfig"
```
Comment (predragnikolic) [inline schemas_extra.json]: I think that you don't need "sublime://schemas/tsconfig" and "sublime://schemas/jsconfig" because they are already defined in `schemas.json` file :)

File: scripts/update_schemas.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3
+
```
Comment (predragnikolic) [inline scripts/update_schemas.py]: What do you think of creating a TextCommand that could be triggered from the command palette. Example `LSP-json: Update schemas` and that command would run this code bellow?

This can be an improvement for later, if you want to. 
But am happy that you took the time to create a script :)

**Comment by jfcherng** [comment]
image

Got this when editing a `.sublime-commands` file.

File: schemas/sublime-commands.json [hunk 11]
```diff
@@ -0,0 +1,7 @@
+{
+  "$id": "sublime://sublime-commands",
+  "title": "Sublime Text Commands",
+  "allowComments": true,
+  "allowTrailingCommas": true,
+  "type": "object"
```
Comment (jfcherng) [inline schemas/sublime-commands.json]: should be `array`

File: schemas/sublime-menu.json [hunk 12]
```diff
@@ -0,0 +1,7 @@
+{
+  "$id": "sublime://sublime-menu",
+  "title": "Sublime Text Menu",
+  "allowComments": true,
+  "allowTrailingCommas": true,
+  "type": "object"
```
Comment (jfcherng) [inline schemas/sublime-menu.json]: should be `array`

File: schemas/sublime-macro.json [hunk 13]
```diff
@@ -0,0 +1,7 @@
+{
+  "$id": "sublime://sublime-macro",
+  "title": "Sublime Text Macro",
+  "allowComments": true,
+  "allowTrailingCommas": true,
+  "type": "object"
```
Comment (jfcherng) [inline schemas/sublime-macro.json]: should be `array`

File: schemas/sublime-mousemap.json [hunk 14]
```diff
@@ -0,0 +1,7 @@
+{
+  "$id": "sublime://sublime-mousemap",
+  "title": "Sublime Text Mousemap",
+  "allowComments": true,
+  "allowTrailingCommas": true,
+  "type": "object"
```
Comment (jfcherng) [inline schemas/sublime-mousemap.json]: should be `array`

File: schemas/sublime-theme.json [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{
+  "$id": "sublime://sublime-theme",
+  "title": "Sublime Text Theme",
+  "allowComments": true,
+  "allowTrailingCommas": true,
+  "type": "object"
```
Comment (jfcherng) [inline schemas/sublime-theme.json]: should be `array`

File: scripts/update_schemas.py (hunk 10 above)
Comment (rchl) [reply scripts/update_schemas.py]: I don't think they change often enough for it to be worth it. Also, since the package is compressed, we would have to save to cache folder. And also we would need to add dependency on requests package also. But my main argument is that I think it's not worth it.

File: schemas/sublime-commands.json (hunk 11 above)
Comment (rchl) [reply schemas/sublime-commands.json]: Thanks, clearly a bad copy paste job :)

File: schemas_extra.json (hunk 9 above)
Comment (rchl) [reply schemas_extra.json]: The ones that are in schemas.json are official specs that abide schema drafts while those here are additional augmentations for those official ones that contain noon-standard properties (vscode extensions). To allow comments, trailing commas and a couple other things you have to use those non-standard properties.
It's ok to provide more than one schema for same file pattern

File: plugin.py (hunk 4 above)
Comment (rchl) [reply plugin.py]: Fixed but since I don't have any automatic checking for those, I'm bound to get this wrong. Is there any way to configure `pyls` to do mypy linting? Or rather, how to do it, as I know that it's supposed to support mypy.
(But then there are bigger errors in this project when trying to run mypy on it)

File: plugin.py (hunk 5 above)
Comment (rchl) [reply plugin.py]: Definitively, we can load schema files from other projects. Although we will need file mapping too so we would need to store that configuration somewhere. VSCode does that - it collects schemas from all installed extensions and stuff, but it's all javascript so it can get those from `package.json`.

File: schemas/jsconfig.json (hunk 6 above)
Comment (rchl) [reply schemas/jsconfig.json]: I hope it's clear from my other response.

File: schemas/vscode-base.json (hunk 8 above)
Comment (rchl) [reply schemas/vscode-base.json]: It's just to enabled comments and trailing commas in project-wide vscode configuration files. Not that relevant for Sublime but I'm working on projects where developers split between vscode and sublime so those are common for me at least.

File: plugin.py (hunk 4 above)
Comment (jfcherng) [reply plugin.py]: > Is there any way to configure `pyls` to do mypy linting?

https://github.com/tomv564/pyls-mypy seems work (with `live_mode` enabled though)

Comment (rwols) [reply plugin.py]: @rchl did you install pyls with `pip3 install python-language-server[all]` ? If you installed it with just `pip3 install python-language-server` (without the `[all]`) then it's missing a few hooks/plugins, among which pyls-mypy.

Comment (jfcherng) [reply plugin.py]: `pyls-mypy` is not include in `python-language-server[all]`

Comment (rchl) [reply plugin.py]: I just can't get it to work.
Installed both with pip3 from brew (`/usr/local/bin/pip3`) and one I have in osx (`/usr/bin/pip3`) but I'm still not sure if language server is seeing it...

Comment (rchl) [reply plugin.py]: OK, got it to work.
`pyls` was started by Python 2 as it was what `/usr/local/bin/python` pointed at. Linking Python 3 to `/usr/local/bin/python` fixed it (almost, also had to `pip3 install future` for `pyls_mypy` plugin).

Comment (jfcherng) [reply plugin.py]: does `pyls_mypy` highlight relative imports as errors on your side?

Comment (rchl) [reply plugin.py]: Sort of... probably not for the same reason as for you as non-relative imports also have similar error:

Screenshot 2020-03-18 at 11 54 15
